File: lib/integrations/matterport.ts
// Matterport API integration for virtual property tours
// Documentation: https://matterport.github.io/showcase-sdk/

export interface MatterportTour {
  id: string
  name: string
  embedUrl: string
  thumbnailUrl: string
  created: string
  propertyAddress?: string
}

export class MatterportService {
  private apiKey: string
  private baseUrl = "https://my.matterport.com/api/v1"

  constructor() {
    this.apiKey = process.env.MATTERPORT_API_KEY || ""
  }

  async getTours(): Promise<MatterportTour[]> {
    try {
      const response = await fetch(`${this.baseUrl}/models`, {
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
        },
      })

      if (!response.ok) {
        throw new Error("Failed to fetch Matterport tours")
      }

      const data = await response.json()
      console.log("[v0] Matterport tours fetched successfully")
      return data.models || []
    } catch (error) {
      console.error("[v0] Matterport API error:", error)
      // Return mock data for demo
      return this.getMockTours()
    }
  }

  async getTourById(id: string): Promise<MatterportTour | null> {
    try {
      const response = await fetch(`${this.baseUrl}/models/${id}`, {
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
        },
      })

      if (!response.ok) {
        throw new Error("Failed to fetch Matterport tour")
      }

      const data = await response.json()
      return data
    } catch (error) {
      console.error("[v0] Matterport API error:", error)
      return null
    }
  }

  generateEmbedUrl(modelId: string): string {
    return `https://my.matterport.com/show/?m=${modelId}`
  }

  private getMockTours(): MatterportTour[] {
    return [
      {
        id: "SxQL3iGyoDo",
        name: "Luxury Villa - Dubai Marina",
        embedUrl: "https://my.matterport.com/show/?m=SxQL3iGyoDo",
        thumbnailUrl: "/luxury-villa-dubai.png",
        created: "2024-01-15",
        propertyAddress: "Dubai Marina, Dubai, UAE",
      },
      {
        id: "j4RZx7ZGM6T",
        name: "Penthouse - Burj Khalifa",
        embedUrl: "https://my.matterport.com/show/?m=j4RZx7ZGM6T",
        thumbnailUrl: "/penthouse-burj-khalifa.jpg",
        created: "2024-02-20",
        propertyAddress: "Downtown Dubai, UAE",
      },
    ]
  }
}

File: backend/app/services/matterport.py
from typing import Any

import httpx

from app.config import settings


class MatterportService:
    base_url = "https://my.matterport.com/api/v1"

    async def get_tours(self) -> list[dict[str, Any]]:
        if not settings.matterport_api_key:
            return self._mock_tours()

        try:
            async with httpx.AsyncClient(timeout=15) as client:
                response = await client.get(
                    f"{self.base_url}/models",
                    headers={"Authorization": f"Bearer {settings.matterport_api_key}"},
                )
                response.raise_for_status()
                return response.json().get("models", [])
        except Exception:
            return self._mock_tours()

    async def get_tour_by_id(self, model_id: str) -> dict[str, Any] | None:
        if not settings.matterport_api_key:
            return next((tour for tour in self._mock_tours() if tour["id"] == model_id), None)

        try:
            async with httpx.AsyncClient(timeout=15) as client:
                response = await client.get(
                    f"{self.base_url}/models/{model_id}",
                    headers={"Authorization": f"Bearer {settings.matterport_api_key}"},
                )
                response.raise_for_status()
                return response.json()
        except Exception:
            return None

    def _mock_tours(self) -> list[dict[str, Any]]:
        return [
            {
                "id": "SxQL3iGyoDo",
                "name": "Luxury Villa - Dubai Marina",
                "embedUrl": "https://my.matterport.com/show/?m=SxQL3iGyoDo",
                "thumbnailUrl": "/luxury-villa-dubai.png",
                "created": "2024-01-15",
                "propertyAddress": "Dubai Marina, Dubai, UAE",
            },
            {
                "id": "j4RZx7ZGM6T",
                "name": "Penthouse - Burj Khalifa",
                "embedUrl": "https://my.matterport.com/show/?m=j4RZx7ZGM6T",
                "thumbnailUrl": "/penthouse-burj-khalifa.jpg",
                "created": "2024-02-20",
                "propertyAddress": "Downtown Dubai, UAE",
            },
        ]
